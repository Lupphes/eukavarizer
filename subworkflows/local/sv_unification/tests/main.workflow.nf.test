nextflow_workflow {

    name "Test Workflow SV_UNIFICATION"
    script "../main.nf"
    workflow "SV_UNIFICATION"
    config "subworkflows/local/sv_unification/tests/nextflow.config"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "sv_unification"

    /*
     * Test scenario 1: Single sample with multiple VCF files for SURVIVOR merge
     * This test validates the SURVIVOR-based merging of multiple SV calls from different callers
     * Expected behavior: Merge VCF files based on breakpoint distance, filter by size, and generate stats
     */
    test("Multiple VCFs - SURVIVOR merge - stub") {
        options '-stub'

        when {
            workflow {
                """
                // Multiple VCF files from different callers for the same sample
                // Using existing test VCF files from nf-core test datasets (gvcf directory for uncompressed VCFs)
                input[0] = channel.of(
                    [[id: 'sample1_cutesv'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf', checkIfExists: true)],
                    [[id: 'sample1_sniffles'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test2.genome.vcf', checkIfExists: true)]
                )
                input[1] = channel.of(
                    [[id: 'sample1_cutesv'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/haplotypecaller_calls/test2_haplotc.vcf.gz', checkIfExists: true)],
                    [[id: 'sample1_sniffles'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/paired_mutect2_calls/test_test2_paired_mutect2_calls.vcf.gz', checkIfExists: true)]
                )
                input[2] = channel.of(
                    [[id: 'sample1_cutesv'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/haplotypecaller_calls/test2_haplotc.vcf.gz.tbi', checkIfExists: true)],
                    [[id: 'sample1_sniffles'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/paired_mutect2_calls/test_test2_paired_mutect2_calls.vcf.gz.tbi', checkIfExists: true)]
                )
                input[3] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.gz', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.survivor_vcf != null },
                { assert workflow.out.survivor_stats != null },
                { assert workflow.out.bcfmerge_vcf != null },
                { assert workflow.out.bcfmerge_tbi != null },
                { assert workflow.out.bcfmerge_stats != null },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    /*
     * Test scenario 2: Single VCF file input
     * This test validates the workflow can handle minimal input (single caller output)
     * Expected behavior: Process single VCF through both SURVIVOR and BCFtools pipelines
     */
    test("Single VCF - minimal input - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = channel.of(
                    [[id: 'sample1'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf', checkIfExists: true)]
                )
                input[1] = channel.of(
                    [[id: 'sample1'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/haplotypecaller_calls/test2_haplotc.vcf.gz', checkIfExists: true)]
                )
                input[2] = channel.of(
                    [[id: 'sample1'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/haplotypecaller_calls/test2_haplotc.vcf.gz.tbi', checkIfExists: true)]
                )
                input[3] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.gz', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.survivor_vcf != null },
                { assert workflow.out.survivor_stats != null },
                { assert workflow.out.bcfmerge_vcf != null },
                { assert workflow.out.bcfmerge_tbi != null },
                { assert workflow.out.bcfmerge_stats != null }
            )
        }
    }

    /*
     * Test scenario 3: Three VCF files to test multiple caller merging
     * This test validates merging of three different VCF files
     * Expected behavior: Merge across multiple VCF files, producing unified output
     * Note: File name collision is avoided by using different test files
     */
    test("Three VCFs - complex merge - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = channel.of(
                    [[id: 'sample1_cutesv'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test.genome.vcf', checkIfExists: true)],
                    [[id: 'sample1_sniffles'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gvcf/test2.genome.vcf', checkIfExists: true)]
                )
                input[1] = channel.of(
                    [[id: 'sample1_cutesv'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/haplotypecaller_calls/test2_haplotc.vcf.gz', checkIfExists: true)],
                    [[id: 'sample1_sniffles'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/paired_mutect2_calls/test_test2_paired_mutect2_calls.vcf.gz', checkIfExists: true)]
                )
                input[2] = channel.of(
                    [[id: 'sample1_cutesv'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/haplotypecaller_calls/test2_haplotc.vcf.gz.tbi', checkIfExists: true)],
                    [[id: 'sample1_sniffles'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/gatk/paired_mutect2_calls/test_test2_paired_mutect2_calls.vcf.gz.tbi', checkIfExists: true)]
                )
                input[3] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.gz', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.survivor_vcf != null },
                { assert workflow.out.survivor_stats != null },
                { assert workflow.out.bcfmerge_vcf != null },
                { assert workflow.out.bcfmerge_tbi != null },
                { assert workflow.out.bcfmerge_stats != null }
            )
        }
    }
}
