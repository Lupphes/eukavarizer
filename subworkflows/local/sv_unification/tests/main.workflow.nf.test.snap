{
    "Multiple VCFs - SURVIVOR merge - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "survivor_merge"
                        },
                        "survivor_merge_survivor_merge_filtered.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "survivor_merge"
                        },
                        "survivor_merge.stats:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "bcftools_concat"
                        },
                        "bcftools_concat.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "bcftools_concat"
                        },
                        "bcftools_concat.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "bcftools_concat"
                        },
                        "bcftools_concat.bcftools_stats.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    "versions.yml:md5,34416b061b5b06e6b88411520f15d279",
                    "versions.yml:md5,582cf2dd9b93657c7766db82e41f9c2d",
                    "versions.yml:md5,64f9431cfda16903407f91347629f098",
                    "versions.yml:md5,6e03f1bde763433bf418503f519e5a5f",
                    "versions.yml:md5,8953b6495de2d66a743e61015eef2c7a",
                    "versions.yml:md5,ab78f9223e2ada690f19086070f6f9e8",
                    "versions.yml:md5,fc7b8a3d99cf981e92a7b570af624d22"
                ],
                "bcfmerge_stats": [
                    [
                        {
                            "id": "bcftools_concat"
                        },
                        "bcftools_concat.bcftools_stats.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bcfmerge_tbi": [
                    [
                        {
                            "id": "bcftools_concat"
                        },
                        "bcftools_concat.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bcfmerge_vcf": [
                    [
                        {
                            "id": "bcftools_concat"
                        },
                        "bcftools_concat.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "survivor_stats": [
                    [
                        {
                            "id": "survivor_merge"
                        },
                        "survivor_merge.stats:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "survivor_vcf": [
                    [
                        {
                            "id": "survivor_merge"
                        },
                        "survivor_merge_survivor_merge_filtered.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,34416b061b5b06e6b88411520f15d279",
                    "versions.yml:md5,582cf2dd9b93657c7766db82e41f9c2d",
                    "versions.yml:md5,64f9431cfda16903407f91347629f098",
                    "versions.yml:md5,6e03f1bde763433bf418503f519e5a5f",
                    "versions.yml:md5,8953b6495de2d66a743e61015eef2c7a",
                    "versions.yml:md5,ab78f9223e2ada690f19086070f6f9e8",
                    "versions.yml:md5,fc7b8a3d99cf981e92a7b570af624d22"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-22T01:37:41.588628996"
    }
}